### PR TITLE
chore: Add dependencies and configuration

### LINKED ISSUE #3: Setup and configure code quality tools
Discuss and agree upon the code quality tools for this project.

I'd recommend the following:
- [Editorconfig](https://editorconfig.org/)
- [pre-commit](https://pre-commit.com/)
- [Black](https://black.readthedocs.io/en/stable/)
- [Isort](https://pycqa.github.io/isort/)
- [Flake8](https://flake8.pycqa.org/en/latest/)
- [Django Upgrade](https://github.com/adamchainz/django-upgrade)
- [Curlylint](https://www.curlylint.org/)
- [DjHTML](https://github.com/rtts/djhtml)

Arno Comment (2023-09-28 05:34):
In the absence of full issues, for now I'll give me feedback this way :D 
YES, I really like the idea of having these code quality things included right from the beginning (in a project where every collaborator is familiar with these tools, that is. I wouldn't recommend to go so heavy with code quality if more novice programmers are involved in the project).

Django Update, CurlyLint and DJHTML I didn't know up to this point. Just checked out their documentation and the seem solid. Very happy to give them a whirl. (Though I mostly use PyCharm and that luckily does a pretty good job with formatting of Django templates & HTML already.)

I created another work item to enable the "Issues" functionality for the repo and assigned it to you.
Also, feel welcome to already set up the configuration files for the above mentioned tools. Even with this initial setup I would appreciate if it takes place in a feature branch, so we can review + approve before merging back into master and get into a clean habit straight away from the very inception. 🚀 

Arno Comment (2023-09-28 05:42):
- I assigned this work item to you and changed the status to "Todo" 😊

### PR DESCRIPTION
**Pull request by michaelwknott**
Decisions for discussion/approval:

**1. I've used the conventional commit specification for this commit message.**
I'm keen to ensure a clean git history and like the conventional commit format. However, I'm open to discussion on what commit message format we should use.
link: https://www.conventionalcommits.org/en/v1.0.0/

**2. Using pip-tools for dependency management**
I like pip-tools as it documents the parent package of each dependency installed in the project. In my opinion this makes dependency management easier in the longer term when adding or removing packages. pip tools also contains a `pip-sync` command that updates the virtual environment to match the `requirements.txt` or `requirements-dev.txt` files, simplifying installing or uninstalling packages.
Link: https://github.com/jazzband/pip-tools

**3. Using separate requirement files for production and development dependencies**
I've created separate requirement.in and requirements-dev.in files and used `pip-compile` to create the repective `requirements.txt` and `requirements-dev.txt` files. I realise this adds 4 files for dependency management which might not be to everyone's taste. As an alternative pip tools provides the option of managing dependencies in a `pyproject.toml` file.

**4. I've assumed we will use Python 3.11**
If changing Python version `.pre-commit-config.yaml` and `pyproject.toml` will require updating:
`.pre-commit-config.yaml` references python 3.11 as the default python version for hooks
`pyproject.toml` references python 3.11 as the target version for black

Resolves #3

### CODE REVIEW

File: .editorconfig
Comment (arnoan) [inline .editorconfig:1]: Perfect! I love that we start off with a set of clear helpful conventions right from the start 💯 
And the the same time keep it basic. The configuration you proposed seems like a perfect balance to me 👍

File: requirements.in
Comment (arnoan) [inline requirements.in:1]: I am very happy for us to use `pip-tools` -> I do the same and recommend it to whoever is open to listen 😄 

That said, I usually include a specific major and minor version for the few "core" packages of the project too. In this case I would use `Django>=4.2,<5`. That would allow us to update other third-party packages we may require (like django-allauth etc.) more regularly by re-running pip-compile. And at the same time keep Django fixed to 4.2 until we want to deliberately update it. (The `Django>=4.2,<5` would still pick up new patch/bugfix releases of Django.)

@michaelwknott 
What do you think?

File: requirements-dev.in [hunk 1]
```diff
@@ -0,0 +1,10 @@
+-c requirements.txt
```
Comment (arnoan) [inline requirements-dev.in:1]: Love it ❤️ I have to say: To me it already feels like you have better practices in place, than the majority of professional developers. Very good! 🥇

File: pyproject.toml [hunk 2]
```diff
@@ -0,0 +1,9 @@
+[tool.black]
+target-version = ['py311']
```
Comment (arnoan) [inline pyproject.toml:2]:
```suggestion
target-version = ['py311']
line-length = 100
```

Can I request for us to use a line length of 100 characters?
Some context: https://knox.codes/posts/line-length-limits

File: pyproject.toml [hunk 3]
```diff
@@ -0,0 +1,9 @@
...
+
+[tool.isort]
+profile = "black"
```
Comment (arnoan) [inline pyproject.toml]:
```suggestion
profile = 'black'
line_length = 100
```

File: pyproject.toml [hunk 4]
```diff
@@ -0,0 +1,9 @@
...
+
+[flake8]
+max-line-length = 88
```
Comment (arnoan) [inline pyproject.toml]:
```suggestion
max-line-length = 100
```

File: pyproject.toml
Comment (arnoan) [inline pyproject.toml:1]: Can we use a character limit of 100 please? 😊

**Review by arnoan — CHANGES_REQUESTED** [review]
@michaelwknott 
Very happy with that base configuration, great work!
I added a few comments (mostly praise 😉) but I would like to request an increase of our default line length limit to 100 chars.

**Comment by arnoan** [comment]
> Decisions for discussion/approval:
> 
> **1. I've used the conventional commit specification for this commit message.** I'm keen to ensure a clean git history and like the conventional commit format. However, I'm open to discussion on what commit message format we should use. link: https://www.conventionalcommits.org/en/v1.0.0/
> 
> **2. Using pip-tools for dependency management** I like pip-tools as it documents the parent package of each dependency installed in the project. In my opinion this makes dependency management easier in the longer term when adding or removing packages. pip tools also contains a `pip-sync` command that updates the virtual environment to match the `requirements.txt` or `requirements-dev.txt` files, simplifying installing or uninstalling packages. Link: https://github.com/jazzband/pip-tools
> 
> **3. Using separate requirement files for production and development dependencies** I've created separate requirement.in and requirements-dev.in files and used `pip-compile` to create the repective `requirements.txt` and `requirements-dev.txt` files. I realise this adds 4 files for dependency management which might not be to everyone's taste. As an alternative pip tools provides the option of managing dependencies in a `pyproject.toml` file.
> 
> **4. I've assumed we will use Python 3.11** If changing Python version `.pre-commit-config.yaml` and `pyproject.toml` will require updating: `.pre-commit-config.yaml` references python 3.11 as the default python version for hooks `pyproject.toml` references python 3.11 as the target version for black
> 
> Resolves #3

Regarding 1️⃣ (commit specification): I never used this conventional specification explicitly before. Just read through the linked resource and like what I see. Explicit is better than implicit and it all feels very pragmatic, sensible and not bloated at all. So YES please, count me in!

2️⃣ & 3️⃣ (pip-compile & separate requirement files): 💯. I use them myself similarly and have no problem with multiple files. If at a later stage we ever want/feel the need for less "clutter" at the project root, we could move them into a dedicated sub-directory and maintain a single `requirements.txt` file at the project root that references them as follows:
```requirements.txt
# Convenience file to reference the other files from a single place
-r requirements/requirements.txt
-r requirements/requirements-dev.txt
```
But that's overkill for now. Let's keep it simple for now, exactly as you proposed.

4️⃣ (Python 3.11): Happy to go with Python 3.11, yes. This coming Monday Python 3.12 will be released, I believe. That said it may take some time for all the packages to be compatible. So I'm very happy to stick with 3.11.

In short: I agree with EVERYTHING 👌

File: requirements.in [hunk 5]
```diff
@@ -0,0 +1 @@
+django
```
Comment (arnoan) [inline requirements.in]:
```suggestion
django>=4.2,<5
```

File: pyproject.toml
Comment (michaelwknott) [reply pyproject.toml:1]: Yes, more than happy to set the character limit to 100 👌

File: requirements-dev.in (hunk 1 above)
Comment (michaelwknott) [reply requirements-dev.in:1]: Thanks Arno

File: requirements.in
Comment (michaelwknott) [reply requirements.in:1]: Perfect, thanks Arno. This makes perfect sense and will allow us to control when/if we upgrade to Django 5. 👌

**Comment by michaelwknott** [comment]
Thanks for the review, Arno. I appreciated the detail you've gone to and the additional resources for context 👍

I've added the requested changes and also moved flake8 configuration from `pyproject.toml` to `.flake8`. I forgot that flake8 doesn't support configuration in `pyproject.toml`.

I'm really enjoying working on this together and it feels great to be collaborating on a project again! 😄

Looking forward to the next steps

**Review by arnoan — APPROVED** [review]
Perfect, thanks for the updates!
From my side this PR is now approved to merge back to `main` ✅

**Comment by michaelwknott** [comment]
Thanks, @arnoan

One question...I'm keen to keep our git history clean. Therefore, as I've pushed two commits with the PR I'm thinking it would be best to 'Squash and Merge' this branch to main and combine the messages into one coherent commit.

Are you happy with the above? Additionally, if you have any other recommendations/best practices I'm happy to take those onboard.

**Comment by arnoan** [comment]
> Thanks, @arnoan
> 
> One question...I'm keen to keep our git history clean. Therefore, as I've pushed two commits with the PR I'm thinking it would be best to 'Squash and Merge' this branch to main and combine the messages into one coherent commit.
> 
> Are you happy with the above? Additionally, if you have any other recommendations/best practices I'm happy to take those onboard.

@michaelwknott 

Mhh.. I have never used a "squash" to be honest. For my projects at work we have a policy in place that only allows for the "basic merge" i.e. adding all commits as they are from the incoming feature branch when performing a merge. The main tradeoff as I see it is whether the priority is on "seeing things as they _really_ happened", or presenting them in a "curated way", possibly to keep the noise levels down. I can see good arguments for both.

At work I am happy we have the choice limited to just one type of merge. Decision fatigue is real 😅 and the act of "squashing" imho introduces one more thing one needs to think about (and presumably also decide how to slice it and what parts of the commit messages of the incoming commits to include and how best to aggregate those). 

That said, please feel VERY welcome to squash here. I'd also be curious to experience what that feels like/looks like. And this project lives under your namespace so you are the BDFL 👑 😉

**Comment by arnoan** [comment]
> Additionally, if you have any other recommendations/best practices I'm happy to take those onboard.

@michaelwknott 

The only thing that currently comes to mind is that I saw that you delegated this PR to me. The practice I am used to is that the person that was primarily responsible for the code included in that PR remains the one "responsible" for it and is also the one that in the end – once all blocking comments (and possibly other pre-merge checks like all unit-tests passing etc.) are addressed – merges the PR.

The other collaborators are usually added only as "reviewers". So I will assign this PR back to you now and also hand you the magic decision wand 🪄 regarding the "to squash or not to squash" conundrum

**Comment by michaelwknott** [comment]
That's useful feedback, thank you. My intention was to assign you as reviewer rather than assign the PR to you. I agree the PR should stay assigned to the original developer. I'll check docs on assigning reviewers 👍🏻

**Comment by arnoan** [comment]
@michaelwknott 
Awesome, thanks for the merge!

**Comment by michaelwknott** [comment]
Hi @arnoan 

I've decided to 'Squash and Merge' this PR to combine commits. I decided on this action as the extra commit related to updating the original PR and added no additional context (it can be considered 'work-in-progress').
